src/ck-core/ckarray.C

```diff
@@ -1743,4 +1778,406 @@ void CkArray::ckDestroy()
+      // a lot of migration, the number of hops is likely to be 2 or less anyways.
+      //if (msg->array_hops() > 1 && CkMyPe() != pe)
+      //{
+      //  pe = locMgr->homePe(id);
+      //}
+      sendToPe(msg, pe, type, opts);
+    }
+  }
+}
+
+void CkArray::recordSend(const CmiUInt8 id, const unsigned int bytes, int pe, const int opts)
+{
+#if CMK_LBDB_ON
+  if (!(opts & CK_MSG_LB_NOTRACE) && locMgr->getLBMgr()->CollectingCommStats())
+  {
+    // LB deals in IDs with collection information only when CMK_GLOBAL_LOCATION_UPDATE
+    // is enabled, so add the group information if so.
+#  if CMK_GLOBAL_LOCATION_UPDATE
+    const CmiUInt8 lbObjId = ck::ObjID(thisgroup, id).getID();
+#  else
+    const CmiUInt8 lbObjId = id;
+#  endif
+    locMgr->getLBMgr()->Send(locMgr->myLBHandle, lbObjId, bytes, pe, 1);
+  }
+#endif
+}
+
+// Send the msg to the designated PE. This is a private entry method and only called by
+// other methods in CkArray when the message is ready to send to the specified PE.
+// If pe is this PE then we know the object is supposed to be located here, but it is
+// possible that it has either been deleted or not been created yet.
+void CkArray::sendToPe(CkArrayMessage* msg, int pe, CkDeliver_t type, int opts)
+{
+  // This method should only be called for a valid PE, and with a properly filled in msg.
+  CkAssert(pe >= 0 && pe < CkNumPes());
+  CkAssert(thisgroup == UsrToEnv(msg)->getArrayMgr());
+  CkAssert(UsrToEnv(msg)->getRecipientID() != 0);
+
+  // If the message is not for me, or is supposed to be queued, send it via the normal
+  // queuing infrastructure
+  if (pe != CkMyPe() || type == CkDeliver_queue)
+  {
+#if CMK_LBDB_ON
+    // Track message sends if the LB framework is collecting comm stats
+    if (msg->array_hops() == 0 && !(opts & CK_MSG_LB_NOTRACE) &&
+        locMgr->getLBMgr()->CollectingCommStats())
+    {
+      recordSend(msg->array_element_id(), UsrToEnv(msg)->getTotalsize(), pe, opts);
+    }
+#endif
+    CkArrayManagerDeliver(pe, msg, opts);
+  }
+  else
+  {
+    // If it is for me, and inline delivery, attempt to invoke the entry method
+    // NOTE: We should only end up here when an inline entry method is called via callback
+    // or when a buffered message is sent from this PE to this PE. Normal inline sends are
+    // handled directly in the .ci file via generated code.
+    CmiUInt8 id = msg->array_element_id();
+    ArrayElement* elem = lookup(id);
+    if (elem == nullptr)
+    {
+      // The element has not yet been created but the location is known. This means the
+      // element is bound to a sibling that has already been created here.
+      if (msg->array_ifNotThere() == CkArray_IfNotThere_buffer)
+      {
+        // Directly buffer. Since we are the location of a sibling, we don't need to make
+        // a location request.
+        bufferedIDMsgs[id].push_back(msg);
+        return;
+      }
+      else
+      {
+        CkLocRec* rec = locMgr->elementNrec(id);
```

Comment:
At the point when `CkArray` needs some awareness of `CkLocRec`, there may be an encapsulation breakdown. For example, at this location, shouldn't `locMgr` offer a direct mechanism for retrieving the index (without going through `CkLocRec`)?

Reply:
Yes, probably. That's a good point. In what was planned to be the next PR in this chain, CkLocRec is removed completely, so I think I just overlooked this piece.

Reply:
Got rid of this dependency and made a bunch of these types of methods private to CkLocMgr.